qark/modules/findPending.py

```diff
@@ -21,456 +22,499 @@
+                                for f in t._fields:
+                                    # dynamically parse each token where is f is the field and t is the token
+                                    try:
+                                        recurse(f, t, results)
+                                    except Exception as e:
+                                        common.logger.debug("Problem in recurse function of findPending.py: " + str(e))
+                                        common.parsingerrors.add(str(current_file))
+
+        except Exception:
+            common.logger.debug("No type declarations: " + str(j))
+            report.write("parsing error-issues-list", str(current_file), "strong")
+    queue.put(results)
+    return
+
+
+def empty_intent(t, filename, results):
+    if type(t) is m.MethodDeclaration:
+        for x in method_list:
+            if str(x) in str(t):
+                fileBody = str(open(filename, 'r').read())
+                for line in fileBody.splitlines():
+                    for y in method_list:
+                        if str(y) in str(line):
+                            if "new Intent()" in str(line):
+                                issue = ReportIssue()
+                                issue.setCategory(ExploitType.CERTIFICATE)
+                                issue.setDetails("Empty Pending instance found in: " + str(
+                                    filename) + ". For security reasons, the Intent you supply here should almost always be an explicit intent that is specify an explicit component to be delivered to through Intent.setClass. A malicious application could potentially intercept, redirect and/or modify this Intent. Pending Intents retain the UID of your application and all related permissions, allowing another application to act as yours. Reference: https://developer.android.com/reference/android/app/PendingIntent.html")
+                                issue.setFile(filename)
+                                issue.setSeverity(Severity.WARNING)
+                                results.append(issue)
+
+                                issue = terminalPrint()
+                                issue.setLevel(Severity.WARNING)
+                                issue.setData("Empty Pending instance found in: " + str(
+                                    filename) + ". For security reasons, the Intent you supply here should almost always be an explicit intent that is specify an explicit component to be delivered to through Intent.setClass. A malicious application could potentially intercept, redirect and/or modify this Intent. Pending Intents retain the UID of your application and all related permissions, allowing another application to act as yours. Reference: https://developer.android.com/reference/android/app/PendingIntent.html")
+                                results.append(issue)
+    return
+
+
+def recurse(f, t, results):
```

Comment:
I'd write a bunch of helper functions for this one. It looks like at a high level, it looks like a switch statement on `type(current)`. Perhaps a separate helper function for each type?